#14 ldap abstraction / Login

NKPmedia opened the pull request:
Added a Ldap orm system with entities
Added a UserService
Added the login feature

Commit:
Added LDAP ORM Bundle

Commit:
Added LDAP Entities with some tests

Commit:
Removed multiple Kernels in the LdapEntityTest
Autoset of ou after dn set in PbnlAccount

Commit:
Added login form
Added LoginController
Added login tests
Added Syfony security components to the User and the UserService
Added ldap SSHA password encode
Added showFlash Template (only with error messages)

Commit:
Removed var_dump

Commit:
Added own ldap orm repo because of missing bug fixes in the old one

Commit:
Updated ldap test database

Commit:
Fixed codeclimate issues

Commit:
Fixed codeclimate issues

Commit:
Replaced variable calls with their getters

jj3l commented on app/Resources/views/includes/base.html.twig:
```diff
@@ -0,0 +1,25 @@
+<html>
+    <head>
+        <meta charset="UTF-8" />
+        <title>{% block title %}Willkommen{% endblock %}</title>
```
Keine Ahnung wie Twig funktioniert. Ist `Willkommen` hier der Default-Text? Frage mich nur, warum der hier auf deutsch und oben auf english steht.

jj3l commented on app/Resources/views/includes/showFlashes.html.twig:
```diff
@@ -0,0 +1,9 @@
+            </div>
+        </div>
+    {% endif %}
+{% endfor %}
```
Gibt es einen Grund, warum die letzte Zeile keinen Umbruch hat? Scheint bei den anderen Dateien auch der Fall zu sein. Kenne das eigentlich so, daß jede Zeile umgebrochen wird (vermmeidet komische Diff-Ansichten).

NKPmedia replied:
Jo stimmt

jj3l commented on app/Resources/views/login/login.html.twig:
```diff
@@ -0,0 +1,44 @@
+{% block body %}
+    <div class="container">
+        <br>
+        <br>
```
Wären die Breaks nicht ein CSS-Thema?

NKPmedia replied:
Jo aber eigentlich braucht man die nicht. Also raus

jj3l commented on app/Resources/views/login/login.html.twig:
```diff
@@ -0,0 +1,44 @@
+        <form class="form-signin" action="{{ path('login') }}" method="post">
+            <h2 class="form-signin-heading">Bitte logge dich ein</h2>
+
+            <label for="username" class="sr-only">Benutzername:</label>
```
Noch mal: Ich habe keine Ahnung von Twig. Aber bestimmt gibt's dort doch auch ein Lokalisierungs-Framework, oder? Falls ja, dann würde ich von vornherein alles Sprachspezifische in eine YML-Datei (oder was auch immer) auslagern. Dutzen/Siezen ist im Zweifel auch ein Lokalisierungsthema.

NKPmedia replied:
Yo habe ich mal eingeführt

jj3l commented on app/Resources/views/login/login.html.twig:
```diff
@@ -0,0 +1,44 @@
+            <h2 class="form-signin-heading">Bitte logge dich ein</h2>
+
+            <label for="username" class="sr-only">Benutzername:</label>
+            <input type="text" class="form-control" id="username" name="_username" value="{{ last_username }}" placeholder="Benutzername"  required autofocus>
```
Ist es Absicht, daß das Label und der Placeholder identisch sind? Falls ja, dann würde ich eins davon weglassen. Falls nein, dann würde ich im Placeholder ein Beispiel nehmen, wie "williwoelfling@pbnl.de" (falls es sich beim Username um die E-Mail-Adresse handelt.

NKPmedia replied:
Das Label wird standartmäsig nicht angezeigt. Darum habe ich sie mal raus genommen.

jj3l commented on app/Resources/views/login/login.html.twig:
```diff
@@ -0,0 +1,44 @@
+            <input type="password" class="form-control" id="password" name="_password" placeholder="Passwort" required>
+            <div class="checkbox">
+                <label>
+                    <input type="checkbox" value="remember-me"> Remember me
```
Hier wieder Englisch...

jj3l commented on app/config/config.yml:
```diff
@@ -7,17 +7,26 @@ imports:
+        uri: "%env(LDAP_DOMAIN)%"
+        use_tls: false
+        bind_dn: "%env(LDAP_USER_DN)%"
+        password: "%env(LDAP_PASSWORD)%"
```
BTW: Wenn hier irgendwo Passwörter auftauchen, dann sind das hoffentlich Beispiele. Falls ja, dann sollte das offensichlich sein, z.B. mit "dies_ist_ein_geheimes_passwort". Oder einfach "secret". Dann weiß jeder, daß da skein echtes Passwort ist. Falls nein: 💥

NKPmedia replied:
Die richtigen Passwörter fügt man über Umgebungsvariablen hinzu. Als Fallback habe ich die Passwörter der Testdatenbank eingetragen, damit Unittest funktionieren. Aber die kann ja jeder wissen, da es die Passwörter für die in dem Ordner ldap-test-database hinterlegte Datenbank sind.

jj3l commented on src/AppBundle/Controller/Login/LoginController.php:
```diff
@@ -0,0 +1,35 @@
+ * User: paul
+ * Date: 03.09.17
+ * Time: 22:31
+ */
```
Hat der Header irgendeine Relevanz? Oder kann man den einfach löschen?

NKPmedia replied:
Ne hat er nicht. Raus

jj3l commented on src/AppBundle/Model/Entity/LDAP/PbnlAccount.php:
```diff
@@ -0,0 +1,369 @@
+    protected $ou;
+
+    /**
+     * the username
```
Am Anfang schreibt man groß ;-)
Unnötige Füllwörter und Artikel ("the") würde ich weglassen. Entwickler möge es gerne kurz und prägnant ;-)
Sinnvoll wäre vielleicht, zu dokumentieren, welcher Zeichenvorat (Groß/Kleibuchstaben mit/ohne Leerzeichen, Umlaute ja/nein) für dieses Attribut erlaubt ist und wofür das Attribut verwendet wird. Zeichenvorat könnte auch als Reg Ex angegeben sein. Inhaltlich ginge es hier um den Nickname oder ersatzweise den Vornamen.

jj3l commented on src/AppBundle/Model/Entity/LDAP/PbnlAccount.php:
```diff
@@ -0,0 +1,369 @@
+    protected $givenName = "";
+
+    /**
+     * also the username (should be the same as givenName)
```
Das ist nur halb richtig, wenn ich mich richtig erinnere. Die `uid` wird doch komplett klein geschrieben und ohne Umlaute, oder?

NKPmedia replied:
Also das LDAP akzeptiert alles.

jj3l commented on src/AppBundle/Model/Entity/LDAP/PbnlAccount.php:
```diff
@@ -0,0 +1,369 @@
+    protected $uid = "";
+
+    /**
+     *  real first name
```
Leerzeichen, am Anfang groß...

jj3l commented on src/AppBundle/Model/Entity/LDAP/PbnlAccount.php:
```diff
@@ -0,0 +1,369 @@
+    protected $cn = "";
+
+    /**
+     *  real second name
```
"last" statt "second" name.

jj3l commented on src/AppBundle/Model/Entity/LDAP/PbnlAccount.php:
```diff
@@ -0,0 +1,369 @@
+    protected $sn = "";
+
+    /**
+     *  the user number (should be unique)
```
> 500 wenn ich mich richtig erinnere.

NKPmedia replied:
Wie 500?

jj3l commented on src/AppBundle/Model/Services/UserService.php:
```diff
@@ -0,0 +1,169 @@
+{
+
+    /**
+     * A Referec to the LdapEntityService to work with the ldap
```
"reference"

jj3l commented on src/AppBundle/Model/Services/UserService.php:
```diff
@@ -0,0 +1,169 @@
+    private $ldapEntityManager;
+
+    /**
+     * Just the logger
```
Aha. Wäre ich nicht drauf gekommen ;-)
Ich glaube, der Kommentar kann weg.

jj3l commented on src/AppBundle/Model/Services/UserService.php:
```diff
@@ -0,0 +1,169 @@
+            );
+        }
+
+        $user = $this->entitiesToUser($ldapPbnlAccount);
```
Warum nicht hier schon `return`?

jj3l commented on src/AppBundle/Model/Services/UserService.php:
```diff
@@ -0,0 +1,169 @@
+    private function entitiesToUser(PbnlAccount $ldapPbnlAccount)
+    {
+        // skip the "{SSHA}"
+        $b64 = substr($ldapPbnlAccount->getUserPassword(), 6);
```
Du kannst Dir den Kommentar sparen, wenn Du `"{SSHA}".length` statt `6` schreibst. Vermeidet gleichzeitig die Magic Number.

jj3l commented on src/AppBundle/Model/Services/UserService.php:
```diff
@@ -0,0 +1,169 @@
+        // the salt (given it is a 8byte one)
+        $salt = substr($b64_dec, -8);
+        // the sha1 part
+        $hashedPassword = substr($b64_dec, 0, 20);
```
Wozu muß hier der Passwort-String auseinandergenommen werden?

NKPmedia replied:
Weil Symfony Salt und Hash getrennt behandelt. Und bei SSHA ist das ja zusammen in einem String.

jj3l commented on src/AppBundle/Model/Services/UserService.php:
```diff
@@ -0,0 +1,169 @@
+        $user->generatePasswordAndSalt($ldapPbnlAccount->getUserPassword());
+        $user->setHomePhoneNumber($ldapPbnlAccount->getTelephoneNumber());
+
+        return $user;
```
Gibt's in PHP soetwas, wie eine [tap-Methode](http://www.engineyard.com/blog/five-ruby-methods-you-should-be-using)?

NKPmedia replied:
Ich verstehe noch nicht ganz, was diese hier bringen soll?

NKPmedia replied:
Ich verstehe noch nicht ganz wofür man die hier braucht?

jj3l commented on src/AppBundle/Model/Services/UserService.php:
```diff
@@ -0,0 +1,169 @@
+use Ucsf\LdapOrmBundle\Ldap\LdapEntityManager;
+use Ucsf\LdapOrmBundle\Repository\Repository;
+
+class UserService implements UserProviderInterface
```
Wäre das nicht ein `UserRepository`?

NKPmedia replied:
jo

jj3l reviewed (commented):
Sieht doch schon mal ganz gut aus!

Commit:
Included suggestion because of code review

Added asserts to the entities
Added translation system and exported texts into yml
Added blank lines at the end of files
Removed headers in files
Changed name of UserService to UserRepository
Some other little fixes

Commit:
Removed assertions that are not important
Added CorruptDataInDatabaseException
The UserRepo will throw it if it loads corrupt data

Commit:
Codeclimate fix

bone4 committed:
Code Review: default LDAP gidNumber is 501

bone4 reviewed (approved):
Sieht gut aus
